pkg/controller/service/service_controller.go

```diff
@@ -720,47 +617,104 @@ func (s *ServiceController) syncService(key string) error {
 
 	// service holds the latest service info from apiserver
 	service, err := s.serviceLister.Services(namespace).Get(name)
-	switch {
-	case errors.IsNotFound(err):
-		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
-		glog.Infof("Service has been deleted %v. Attempting to cleanup load balancer resources", key)
-		err = s.processServiceDeletion(key)
-	case err != nil:
-		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
-	default:
-		cachedService = s.cache.getOrCreate(key)
-		err = s.processServiceUpdate(cachedService, service, key)
+	if errors.IsNotFound(err) {
+		glog.V(4).Infof("Service %q has been deleted", key)
+		return nil
 	}
 
-	return err
+	if err != nil {
+		glog.Errorf("Unable to retrieve service %v from store: %v", key, err)
+		return err
+	}
+
+	if isDeletionCandidate(service) {
+		err := s.processServiceDeletion(service, key)
+		if err != nil {
+			return err
+		}
+
+		service, err = s.removeFinalizer(service)
+		if err != nil {
+			return err
+		}
+
+		// The service and all external resources, if applicable, have been removed.
+		return nil
+	}
+
+	if needToAddFinalizer(service) {
+		service, err = s.addFinalizer(service)
+		if err != nil {
+			return err
+		}
+
+		// Continue processing the service now that it has been updated with the finalizer.
+	}
+
+	return s.processServiceUpdate(service, key)
 }
 
 // Returns an error if processing the service deletion failed, along with a time.Duration
 // indicating whether processing should be retried; zero means no-retry; otherwise
 // we should retry after that Duration.
-func (s *ServiceController) processServiceDeletion(key string) error {
-	cachedService, ok := s.cache.get(key)
-	if !ok {
-		glog.Errorf("service %s not in cache even though the watcher thought it was. Ignoring the deletion", key)
-		return nil
-	}
-	return s.processLoadBalancerDelete(cachedService, key)
+func (s *ServiceController) processServiceDeletion(service *v1.Service, key string) error {
+	return s.processLoadBalancerDelete(service, key)
 }
 
-func (s *ServiceController) processLoadBalancerDelete(cachedService *cachedService, key string) error {
-	service := cachedService.state
-	// delete load balancer info only if the service type is LoadBalancer
-	if !wantsLoadBalancer(service) {
```

Comment:
Curious why is this check no longer needed.

Reply:
Because we are now passing in the actual service instead of the cached one. The actual service may already be updated e.g. from `type=LoadBalancer` to `type=ClusterIP`, so the check here no longer makes sense. Instead, I think we now decide whether to process deletion based on having finalizer or not.